File: Schema-Tool-ViewModel/src/main/java/net/bi4vmr/study/base/MyViewModel.java
package net.bi4vmr.study.base;

import android.util.Log;

import androidx.lifecycle.ViewModel;

import java.util.UUID;

/**
 * Name        : MyViewModel
 * <p>
 * Author      : BI4VMR
 * <p>
 * Email       : bi4vmr@outlook.com
 * <p>
 * Date        : 2023-05-11 21:10
 * <p>
 * Description : 测试用ViewModel
 */
public class MyViewModel extends ViewModel {

    private String name;

    public MyViewModel() {
        name = genRandomID();
        Log.i("myapp", "MyViewModel-VM created. Name:" + name);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        Log.i("myapp", "MyViewModel-OnCleared. Name:" + name);
    }

    public String getVMName() {
        return name;
    }

    // 获取随机ID
    private String genRandomID() {
        return UUID.randomUUID()
                .toString()
                .toUpperCase()
                .substring(0, 6);
    }
}

File: M03_UI/C10_Compose/S01_Base/src/main/kotlin/net/bi4vmr/study/data/MyViewModel.kt
package net.bi4vmr.study.data

import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

/**
 * ViewModel：计数器示例。
 *
 * @author bi4vmr@outlook.com
 * @since 1.0.0
 */
class MyViewModel : ViewModel() {

    // 可变状态Flow，仅内部使用。
    private val _state: MutableStateFlow<Int> = MutableStateFlow(0)

    // 不可变状态Flow，与可变实例保持同步，供外部观察。
    val state: StateFlow<Int> = _state


    // 业务接口：数值增加。
    fun add() {
        _state.value++
    }

    // 业务接口：数值减少。
    fun reduce() {
        _state.value--
    }
}
